engine/src/main/java/org/terasology/engine/core/module/ClasspathCompromisingModuleFactory.java

```diff
@@ -52,6 +63,97 @@ public Module createArchiveModule(ModuleMetadata metadata, File archive) throws
                 new ClassesInModule(module));
     }
 
+    /**
+     * Find the location of the module containing this URL.
+     * <p>
+     * Accounts for loading modules from development workspaces that may have their build directories
+     * on the classpath, as is the case when running tests.
+     *
+     * @see #setDefaultCodeSubpath
+     * @see #setDefaultLibsSubpath
+     *
+     * @param metadataName the expected name of the metadata file, as it would appear in {@link #getModuleMetadataLoaderMap()}
+     * @param metadataUrl a URL of a metadata file, such as might be returned from {@link ClassLoader#getSystemResource}
+     * @return the module's base directory, or a jar file if it doesn't look like a local build
+     */
+    Path canonicalModuleLocation(String metadataName, URL metadataUrl) {
+        checkArgument(getModuleMetadataLoaderMap().containsKey(metadataName),
+                "metadataName `%s` is not in loader map", metadataName);
+        if (metadataUrl.getProtocol().equals("jar")) {
+            return modulePathFromMetadataJarUrl(metadataUrl);
+        } else {
+            return modulePathFromMetadataFileUrl(metadataName, metadataUrl);
+        }
+    }
+
+    private Path modulePathFromMetadataFileUrl(String metadata_name, URL url) {
+        Path path = fromUrl(url);
+        // We are considering the location of a resource file, so compare it to the code path.
+        // Include the metadata_name in case it has path components of its own.
+        Path relativePathFromModuleRoot = Paths.get(getDefaultCodeSubpath(), metadata_name);
+        return findModuleRoot(relativePathFromModuleRoot, path).orElse(path);
+    }
```

Comment:
```suggestion
    private Path modulePathFromMetadataFileUrl(String metadataName, URL url) {
        Path path = fromUrl(url);
        // We are considering the location of a resource file, so compare it to the code path.
        // Include the metadata name in case it has path components of its own.
        Path relativePathFromModuleRoot = Paths.get(getDefaultCodeSubpath(), metadataName);
        return findModuleRoot(relativePathFromModuleRoot, path).orElse(path);
    }
```